Exception when formatting code

I just installed the 3.0 release of the plugin

Config went well. The first thing I did is deleting an unused import (the whole line), and an exception occured:

```
com.intellij.execution.process.ProcessNotCreatedException: Cannot run program "/Users/peal6230/.npm-packages/bin": error=13, Permission denied
	at com.intellij.execution.configurations.GeneralCommandLine.createProcess(GeneralCommandLine.java:416)
	at org.elm.openapiext.SubprocessesKt.execute(Subprocesses.kt:103)
	at org.elm.openapiext.SubprocessesKt.execute$default(Subprocesses.kt:97)
	at org.elm.workspace.commandLineTools.ElmFormatCLI.getFormattedContentOfDocument(ElmFormatCLI.kt:37)
	at org.elm.workspace.commandLineTools.ElmFormatCLI.access$getFormattedContentOfDocument(ElmFormatCLI.kt:26)
	at org.elm.workspace.commandLineTools.ElmFormatCLI$formatDocumentAndSetText$result$1.compute(ElmFormatCLI.kt:44)
	at org.elm.workspace.commandLineTools.ElmFormatCLI$formatDocumentAndSetText$result$1.compute(ElmFormatCLI.kt:26)
	at com.intellij.openapi.progress.impl.CoreProgressManager$1.run(CoreProgressManager.java:220)
	at com.intellij.openapi.progress.impl.CoreProgressManager$TaskRunnable.run(CoreProgressManager.java:727)
	at com.intellij.openapi.progress.impl.CoreProgressManager$5.run(CoreProgressManager.java:442)
	at com.intellij.openapi.progress.impl.CoreProgressManager.lambda$runProcess$2(CoreProgressManager.java:164)
	at com.intellij.openapi.progress.impl.CoreProgressManager.registerIndicatorAndRun(CoreProgressManager.java:582)
	at com.intellij.openapi.progress.impl.CoreProgressManager.executeProcessUnderProgress(CoreProgressManager.java:532)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.executeProcessUnderProgress(ProgressManagerImpl.java:87)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcess(CoreProgressManager.java:151)
	at com.intellij.openapi.application.impl.ApplicationImpl.runProcessWithProgressSynchronously(ApplicationImpl.java:571)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcessWithProgressSynchronously(CoreProgressManager.java:454)
	at com.intellij.openapi.progress.impl.ProgressManagerImpl.runProcessWithProgressSynchronously(ProgressManagerImpl.java:110)
	at com.intellij.openapi.progress.impl.CoreProgressManager.runProcessWithProgressSynchronously(CoreProgressManager.java:216)
	at org.elm.workspace.commandLineTools.ElmFormatCLI.formatDocumentAndSetText(ElmFormatCLI.kt:43)
	at org.elm.ide.components.ElmFormatOnFileSaveComponent$initComponent$1.beforeDocumentSaving(ElmFormatOnFileSaveComponent.kt:39)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.util.messages.impl.MessageBusConnectionImpl.deliverMessage(MessageBusConnectionImpl.java:117)
	at com.intellij.util.messages.impl.MessageBusImpl.doPumpMessages(MessageBusImpl.java:397)
	at com.intellij.util.messages.impl.MessageBusImpl.pumpWaitingBuses(MessageBusImpl.java:358)
	at com.intellij.util.messages.impl.MessageBusImpl.pumpMessages(MessageBusImpl.java:347)
	at com.intellij.util.messages.impl.MessageBusImpl.sendMessage(MessageBusImpl.java:324)
	at com.intellij.util.messages.impl.MessageBusImpl.access$200(MessageBusImpl.java:29)
	at com.intellij.util.messages.impl.MessageBusImpl$2.invoke(MessageBusImpl.java:196)
	at com.sun.proxy.$Proxy35.beforeDocumentSaving(Unknown Source)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.intellij.openapi.fileEditor.impl.FileDocumentManagerImpl.multiCast(FileDocumentManagerImpl.java:140)
	at com.intellij.openapi.fileEditor.impl.FileDocumentManagerImpl.lambda$new$0(FileDocumentManagerImpl.java:120)
	at com.sun.proxy.$Proxy35.beforeDocumentSaving(Unknown Source)
	at com.intellij.openapi.fileEditor.impl.FileDocumentManagerImpl.lambda$doSaveDocumentInWriteAction$4(FileDocumentManagerImpl.java:422)
	at com.intellij.pom.core.impl.PomModelImpl.guardPsiModificationsIn(PomModelImpl.java:402)
	at com.intellij.openapi.fileEditor.impl.FileDocumentManagerImpl.doSaveDocumentInWriteAction(FileDocumentManagerImpl.java:421)
	at com.intellij.openapi.fileEditor.impl.FileDocumentManagerImpl.lambda$doSaveDocument$2(FileDocumentManagerImpl.java:383)
	at com.intellij.openapi.application.WriteAction.run(WriteAction.java:106)
	at com.intellij.openapi.fileEditor.impl.FileDocumentManagerImpl.doSaveDocument(FileDocumentManagerImpl.java:383)
	at com.intellij.openapi.fileEditor.impl.FileDocumentManagerImpl.saveAllDocuments(FileDocumentManagerImpl.java:303)
	at com.intellij.openapi.fileEditor.impl.FileDocumentManagerImpl.saveAllDocuments(FileDocumentManagerImpl.java:281)
	at com.intellij.openapi.components.impl.stores.StoreUtil.saveDocumentsAndProjectsAndApp(StoreUtil.java:105)
	at com.intellij.ide.actions.SaveAllAction.actionPerformed(SaveAllAction.java:23)
	at com.intellij.openapi.actionSystem.ex.ActionUtil$1.run(ActionUtil.java:258)
	at com.intellij.openapi.actionSystem.ex.ActionUtil.performActionDumbAware(ActionUtil.java:275)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher$1.performAction(IdeKeyEventDispatcher.java:618)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.lambda$processAction$2(IdeKeyEventDispatcher.java:667)
	at com.intellij.openapi.application.TransactionGuardImpl.performUserActivity(TransactionGuardImpl.java:195)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.processAction(IdeKeyEventDispatcher.java:666)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.processActionOrWaitSecondStroke(IdeKeyEventDispatcher.java:519)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.inInitState(IdeKeyEventDispatcher.java:474)
	at com.intellij.openapi.keymap.impl.IdeKeyEventDispatcher.dispatchKeyEvent(IdeKeyEventDispatcher.java:211)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.java:643)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.java:363)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
Caused by: java.io.IOException: Cannot run program "/Users/peal6230/.npm-packages/bin": error=13, Permission denied
	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1048)
	at com.intellij.execution.configurations.GeneralCommandLine.startProcess(GeneralCommandLine.java:444)
	at com.intellij.execution.configurations.GeneralCommandLine.createProcess(GeneralCommandLine.java:412)
	... 66 more
Caused by: java.io.IOException: error=13, Permission denied
	at java.lang.UNIXProcess.forkAndExec(Native Method)
	at java.lang.UNIXProcess.<init>(UNIXProcess.java:247)
	at java.lang.ProcessImpl.start(ProcessImpl.java:134)
	at java.lang.ProcessBuilder.start(ProcessBuilder.java:1029)
	... 68 more
```

I cancelled the modification, restarted the IDE, deleted the unused import again, and the error occurred again.

The line I deleted is an import from elm-css:

```
import Html.Styled exposing (toUnstyled)
```

--- a/src/main/kotlin/org/elm/ide/actions/ElmExternalFormatAction.kt
+++ b/src/main/kotlin/org/elm/ide/actions/ElmExternalFormatAction.kt
@@ -1,7 +1,6 @@
 package org.elm.ide.actions
 
 import com.intellij.execution.ExecutionException
-import com.intellij.notification.NotificationAction
 import com.intellij.notification.NotificationType
 import com.intellij.openapi.actionSystem.AnAction
 import com.intellij.openapi.actionSystem.AnActionEvent
@@ -36,13 +35,11 @@ class ElmExternalFormatAction : AnAction() {
             return
         }
 
+        val fixAction = "Fix" to { ctx.project.elmWorkspace.showConfigureToolchainUI() }
+
         val elmFormat = ctx.project.elmToolchain.elmFormatCLI
         if (elmFormat == null) {
-            ctx.project.showBalloon("Could not find elm-format", NotificationType.ERROR,
-                    action = NotificationAction.create("Fix") { _, notification ->
-                        notification.hideBalloon()
-                        ctx.project.elmWorkspace.showConfigureToolchainUI()
-                    })
+            ctx.project.showBalloon("Could not find elm-format", NotificationType.ERROR, fixAction)
             return
         }
 
@@ -57,7 +54,7 @@ class ElmExternalFormatAction : AnAction() {
         } catch (ex: ExecutionException) {
             if (isUnitTestMode) throw ex
             val message = ex.message ?: "something went wrong running elm-format"
-            ctx.project.showBalloon(message, NotificationType.ERROR)
+            ctx.project.showBalloon(message, NotificationType.ERROR, fixAction)
         }
     }
 

--- a/src/main/kotlin/org/elm/ide/components/ElmFormatOnFileSaveComponent.kt
+++ b/src/main/kotlin/org/elm/ide/components/ElmFormatOnFileSaveComponent.kt
@@ -1,6 +1,9 @@
 package org.elm.ide.components
 
 import com.intellij.AppTopics
+import com.intellij.execution.ExecutionException
+import com.intellij.execution.process.ProcessNotCreatedException
+import com.intellij.notification.NotificationType
 import com.intellij.openapi.application.ApplicationManager
 import com.intellij.openapi.components.ProjectComponent
 import com.intellij.openapi.editor.Document
@@ -9,10 +12,13 @@ import com.intellij.openapi.fileEditor.FileDocumentManagerListener
 import com.intellij.openapi.project.Project
 import com.intellij.psi.PsiDocumentManager
 import com.intellij.psi.util.PsiTreeUtil
+import org.elm.ide.notifications.showBalloon
 import org.elm.lang.core.psi.isElmFile
+import org.elm.openapiext.isUnitTestMode
 import org.elm.workspace.commandLineTools.ElmFormatCLI
 import org.elm.workspace.elmSettings
 import org.elm.workspace.elmToolchain
+import org.elm.workspace.elmWorkspace
 
 class ElmFormatOnFileSaveComponent(val project: Project) : ProjectComponent {
 
@@ -36,7 +42,18 @@ class ElmFormatOnFileSaveComponent(val project: Project) : ProjectComponent {
                         val elmVersion = ElmFormatCLI.getElmVersion(project, vFile) ?: return
                         val elmFormat = project.elmToolchain.elmFormatCLI ?: return
 
-                        elmFormat.formatDocumentAndSetText(project, document, elmVersion, addToUndoStack = false)
+                        try {
+                            elmFormat.formatDocumentAndSetText(project, document, elmVersion, addToUndoStack = false)
+                        } catch (e: ExecutionException) {
+                            if (isUnitTestMode) throw e
+                            val message = e.message ?: "Something went wrong running elm-format"
+                            val actions = when (e) {
+                                is ProcessNotCreatedException ->
+                                    arrayOf("Fix Path" to { project.elmWorkspace.showConfigureToolchainUI() })
+                                else -> emptyArray()
+                            }
+                            project.showBalloon(message, NotificationType.ERROR, *actions)
+                        }
                     }
                 }
         )

--- a/src/main/kotlin/org/elm/ide/notifications/Utils.kt
+++ b/src/main/kotlin/org/elm/ide/notifications/Utils.kt
@@ -15,10 +15,29 @@ import com.intellij.openapi.project.Project
 
 private val pluginNotifications = NotificationGroup.balloonGroup("Elm Plugin")
 
-fun Project.showBalloon(content: String, type: NotificationType, action: NotificationAction? = null) {
+
+/**
+ * Show a balloon notification along with action(s). The notification will be automatically dismissed
+ * when an action is invoked.
+ *
+ * @param content The main content to be shown in the notification
+ * @param type The notification type
+ * @param actions Optional list of actions to be included in the notification
+ */
+fun Project.showBalloon(
+        content: String,
+        type: NotificationType,
+        vararg actions: Pair<String, (() -> Unit)>
+) {
     val notification = pluginNotifications.createNotification(content, type)
-    if (action != null) {
-        notification.addAction(action)
+    for ((actionTitle, actionCallback) in actions) {
+        notification.addAction(
+                NotificationAction.create(actionTitle)
+                { _, notif ->
+                    notif.hideBalloon()
+                    actionCallback()
+                }
+        )
     }
     Notifications.Bus.notify(notification, this)
 }

